CORDA-4103 Feature Branch: Serialization injection for transaction building

williamvigorr3 opened the pull request.

williamvigorr3 committed:
CORDA-4105 Add public API to allow custom serialization schemes (#6848)

* CORDA-4105 Add public API to allow custom serialization schemes

* Fix Detekt

* Suppress warning

* Fix usused import

* Improve API to use generics

This does not break Java support (only Intelij gets confused).

* Add more detailed documentation to public interfaces

* Change internal variable name after rename public API

* Update Public API to use ByteSquence instead of SerializedBytes

* Remove unused import

* Fix whitespace.

* Add added public API to .ci/api-current.txt

* Improve public interfaces

Rename CustomSchemeContext to SerializationSchemeContext to improve
clarity and move to it's own file. Improve kdoc to make things less
confusing.

* Update API current with changed API

williamvigorr3 committed:
CORDA-4104 Implement custom serialization scheme discovery mechanism (#6854)

* CORDA-4104 Implement CustomSerializationScheme Discovery Mechanism

Discovers a single CustomSerializationScheme implementation inside
the drivers dir using a system property.

* Started MockNetwork test

* Add driver test of Custom Serialization Scheme

* Fix detekt and minor style error

* Respond to review comments

Allow non-single arg constructors (there must be one no args
constructor), move code from SerializationEnviroment into its
own file, improve exceptions to be more user friendly.

* Fix minor bug in Scheme finding code  + improve error messages

williamvigorr3 committed:
CORDA-4104 Improve test coverage of custom serialization scheme discovery (#6855)

* CORDA-4104 Add test of classloader scanning for CustomSerializationSchemes

* Fix Detekt

williamvigorr3 committed:
NOTICK Clarify KDOC on SerializationSchemeContext (#6865)

williamvigorr3 committed:
CORDA-4111 Change Component Group Serialization to use contex when the lazy map is constructed (#6856)

Currently the component group will recheck the thread local (global)
serialization context when component groups are serialized lazily.
Instead store the serialization context when the lazy map is constructed
and use that latter when doing serialization lazily.

williamvigorr3 committed:
CORDA-4106 Test wire transaction can still be written to the ledger (#6860)

* Add test that writes transaction to the Database

* Improve test check serialization scheme in test body

mvdbos commented on core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt:
```diff
@@ -140,6 +143,23 @@ open class TransactionBuilder(
+     * @throws [ZoneVersionTooLowException] if there are reference states and the zone minimum platform version is less than 4.
+     */
+    @Throws(MissingContractAttachments::class)
+    fun toWireTransaction(services: ServicesForResolution, schemeId: Int): WireTransaction {
```
Now that we are adding a new version of this method, can we perhaps make it optionally possible to also provide the serializationContext, and otherwise default to what is on line 159? 

At WireTransaction build time, we would like to pass additional information to the serializer through the context, but for that, we need to be able to choose a custom context too. `SerializationContext.properties` (Map<Any, Any>) seems suitable for that.

This means that this property needs to be also propagated to `SerializationSchemeContext`, which now only copies two of the properties (which makes sense mostly, because those seem specific to Corda AMQP serializer).

williamvigorr3 replied:
We can potentially do this. However you should be aware that you will likely not be able to control the `SerializationContext`/`SerializationSchemeContext` when deserialising (as this is done internally in the node). So if you need the properties when deserializing they might not be in the map.

mvdbos replied:
Wouldn't we be able to call `SerializationContextImpl.withProperty(property: Any, value: Any): SerializationContext` just before calling `TransactionBuilder.toWireTransaction(services: ServicesForResolution, schemeId: Int, serializationContext: SerializationContext): WireTransaction` with it?

mvdbos replied:
Oh wait, you are referring to *de*serialization. Yes, that would be a problem.

mvdbos commented on core/src/main/kotlin/net/corda/core/serialization/SerializationSchemeContext.kt:
```diff
@@ -0,0 +1,23 @@
+     * deserializing. To catch classes missing from the whitelist as early as possible it is HIGHLY recommended to also check this
+     * whitelist when serializing (as well as deserialising) objects.
+     */
+    val whitelist: ClassWhitelist
```
As mentioned in [this](https://github.com/corda/corda/pull/6867/files#r573022017) comment below, it would be great if we can add `val properties: Map<Any, Any>` and have it also copied on construction of `SerializationSchemeAdapter`.

williamvigorr3 committed:
CORDA-4119 Minor changes to serialisation injection for transaction building (#6868)

* CORDA-4119 Minor changes to serialisation injection for transaction building

Scan the CorDapp classloader instead of the drivers classloader.
Add properties map to CustomSerialiaztionContext (copied from SerializationContext).
Change API to let a user pass in the serialization context in TransactionBuilder.toLedgerTransaction

* Improve KDOC + fix shawdowing issue in CordaUtils

* Pass only the properties map into theTransactionBuilder.toWireTransaction

Not the entire serializationContext

* Revert change to CordaUtils

* Improve KDOC explain pitfalls of setting properties
